content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+---
+title: "Towards Continuous Performance Regression Testing"
+date: 2020-12-16T17:45:00+01:00
+draft: false
+---
+:source-highlighter: rouge
+:rouge-style: base16.dark
+:icons: font
+:imagesdir: /images
+ifdef::env-github[]
+:imagesdir: ../../static/images
+endif::[]
+
+Functional unit and integration tests are a standard tool of any software development organization,
+helping not only to ensure correctness of newly implemented code,
+but also to identify regressions -- bugs to existing functionality introduced by a code change.
+The situation looks different though when it comes to regressions related to non-functional requirements, in particular performance-related ones:
+How to detect increased response times in a web application?
+How to identify decreased throughput?
+
+These aspects are typically hard to test in an automated and reliable way in the development workflow,
+as they are dependent on the underlying hardware and the workload of an application.
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
+Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation, database I/O, or number of executed SQL statements, for a given workload.
+Starting off from a defined base line, future failures of such assertions are indicators for potential performance regressions in an application, as a code change may have introduced higher GC pressure,
```

Comment:
I don't think there's such a thing as "GC pressure". you probably mean excessive allocations?

What I'd see your tool most useful for is to introduce regression tests for specific optimisations which have already been identified in the past as problematic under certain workloads. For example, as you well know, there's some code areas of Hibernate ORM which have been carefully tuned for being very lean on allocations, while in 90% of the codebase this doesn't matter (or its importance has not been identified yet).

When one implements such an optimisation, it requires a specialized benchmark to verify that the fix has been performed, but we don't have the means to protect the feature against future regressions.

In such a scenario I'd think JfrUnit would be "gold", as one has a deep understanding of the specific metrics which matter in that case, and can encode the goal as "indirect metrics" to be tested more easily in the future.

Of course that doesn't guarantee any figure to be delivered on the same benchmark again, but it's much better than not doing anything at all.

Reply:
Yes, that's exactly the use case I have in mind. I think you would benefit a lot from JfrUnit in Hibernate by using it to set up a performance regression test suite. But beyond that, I think it also can be interesting for application developers, which is what I'm focusing a bit more in this post (solely due to the fact that that this is a much larger audience).

Reply:
> I don't think there's such a thing as "GC pressure". you probably mean excessive allocations?

The latter is leading to the former :) Seeing quite a few usages of that term, including JDK docs, so I think it's legit.